### PR TITLE
REF: de-duplicate code in libperiod

### PR DESCRIPTION
**Pull request by jbrockmendel**
discussed a few months ago, this gets rid of `date_info_from_days_and_time` and instead re-uses `pandas_datetime_to_datetimestruct`

### CODE REVIEW

**Review by gfyoung — APPROVED** [review]

File: pandas/_libs/tslibs/period.pyx [hunk 1]
```diff
@@ -855,74 +853,49 @@ cdef int64_t get_unix_date(int64_t period_ordinal, int freq) nogil:
...
-    dts.hour = hour
-    dts.min = minute
-    dts.sec = <int>second
+    if freq == FR_NS:
+        factor = 1
 
-    subsecond_fraction = second - dts.sec
-    dts.us = int((subsecond_fraction) * 1e6)
-    dts.ps = int(((subsecond_fraction) * 1e6 - dts.us) * 1e6)
+    if freq == FR_US:
```
Comment (jreback) [inline pandas/_libs/tslibs/period.pyx:882]: elif on these?

**Comment by jbrockmendel** [comment]
cc @WillAyd trying to figure out which of the current CI failures are false-positives, is this one from Travis a real problem?

```
pandas/_libs/tslibs/period.c:9227:11: error: ‘__pyx_v_factor’ may be used uninitialized in this function [-Werror=maybe-uninitialized]

   __pyx_r = (__pyx_v_sub * __pyx_v_factor);

           ^

pandas/_libs/tslibs/period.c:8995:26: note: ‘__pyx_v_factor’ was declared here

   __pyx_t_5numpy_int64_t __pyx_v_factor;

                          ^

cc1: all warnings being treated as errors

building 'pandas._libs.tslibs.resolution' extension
```

**Comment by WillAyd** [comment]
I'm assuming your latest commit fixes the issue, but yea I think just telling you that `factor` could potentially be uninitialized if none of the conditions branch, so either need to guarantee it branches or initialize to a value

File: pandas/_libs/tslibs/period.pyx [hunk 2]
```diff
@@ -806,24 +806,22 @@ cdef int64_t get_period_ordinal(npy_datetimestruct *dts, int freq) nogil:
...
+    nanos = get_time_nanos(freq, unix_date, ordinal)
 
-    while abstime >= 86400:
-        abstime -= 86400
-        unix_date += 1
+    pandas_datetime_to_datetimestruct(unix_date, NPY_FR_D, dts)
 
-    date_info_from_days_and_time(dts, unix_date, abstime)
+    dt64_to_dtstruct(nanos, &dts2)
+    dts.hour = dts2.hour
```
Comment (WillAyd) [inline pandas/_libs/tslibs/period.pyx:820]: Any reason you don't just pass the reference to dts in the call above?

Comment (jbrockmendel) [reply pandas/_libs/tslibs/period.pyx:820]: Because the caller might still need `dts`

Comment (WillAyd) [reply pandas/_libs/tslibs/period.pyx:820]: Doesn't calling dt64_to_dtstruct do exactly what the next 5 lines of code do anyway?

Comment (jbrockmendel) [reply pandas/_libs/tslibs/period.pyx:820]: i can confirm that just passing `dts` does Break The World.  IIRC part of whats going on here is that `dts` may be out of bounds for dt64[ns]

Comment (WillAyd) [reply pandas/_libs/tslibs/period.pyx:820]: Ah OK. Thanks for checking - seems strange to have to do this but probably a separate issue then

File: pandas/_libs/tslibs/period.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/period.pyx]: updated+green

**Pull request merged**